1-Draft/RFCnnnn-ForEach-Parallel-Cmdlet.md

```diff
@@ -0,0 +1,99 @@
+---
+RFC: RFCnnnn
+Author: Paul Higinbotham
+Status: Draft
+SupercededBy: N/A
+Version: 1.0
+Area: Engine
+Comments Due: July 18, 2019
+Plan to implement: Yes
+---
+
+# PowerShell ForEach-Object -Parallel Cmdlet
+
+This RFC proposes a new parameter set for the existing ForEach-Object cmdlet to parallelize script block executions, instead of running them sequentially as it does now.  
+
+## Motivation
+
+    As a PowerShell User,
+    I can do simple fan-out concurrency with the PowerShell ForEach-Object cmdlet, without having to obtain and load a separate module, or deal with PowerShell jobs unless I want to.

```

Comment:
I'm wondering what is meant by "simple" here.  It seems to me that "simple" from the user's perspective means something like "works as I'd expect in common use cases".  Vaguely speaking, I would suggest that such behavior would include the following:

1. Speedup is achieved to the degree that the workload in the scriptblock allows.
2. Scriptblocks can safely be passed as variables to the command.

I'm highlighting these behaviors in particular because they seem like obvious goals yet require overcoming substantial underlying implementation challenges to achieve.  In particular,

1. PowerShell/PowerShell#7377, [#7153](https://github.com/PowerShell/PowerShell/issues/7135), [#7131](https://github.com/PowerShell/PowerShell/issues/7131), and [#7035](https://github.com/PowerShell/PowerShell/issues/7035) all relate to various roadblocks to opening, resetting, and reusing, runspaces in parallel.  The summary is that each runspace will need to load the modules it needs and there currently does not seem to be a way to open modules in parallel in a manner that results in speedup.
2. PowerShell/PowerShell#7626 demonstrates that PowerShell currently can crash (with, _I think_, possibly undefined behavior) when scriptblocks are passed amongst runspaces.

I think overcoming these challenges would be great for PowerShell, but I don't think they should be underestimated.

Reply:
Yes, scriptblocks have runspace affinity and we need resolve this before we can go any further.